explain how to create a machine

marcosh opened the pull request:
I'm pointing this PR not to `main` because it refers to concepts which are not in `main` yet.

The idea of what I wrote so far is to explain all the bits and pieces which one needs to know to create a `StateMachine a b`, including the meaning, the type and the kind of all the parameters used around.

So far, I just did the easy bit. The hard part is to describe how to use properly the `BaseMachine` type. In the `spec/CRM/Example` folder there are several examples which hopefully illustrate how to use such a type. More examples, more complex and illustrative ones, are coming soon.

closes #25

AlexeiDrake commented on docs/how-to-create-a-machine.md:
````diff
@@ -0,0 +1,167 @@
+
+to get a single machine
+
+```
+  ┌─────────────────┐
+a │ ┌╌╌╌╌┐ b ┌╌╌╌╌┐ │ c
+──┼╌┤    ├╌╌╌┤    ├╌┼──
+  │ └╌╌╌╌┘   └╌╌╌╌┘ │
+  └─────────────────┘
+```
````
I certainly appreciate the effort that must have gone into drawing this intricate ASCII art. :laughing: 

Consider using [Mermaid](https://mermaid.js.org/syntax/stateDiagram.html) instead.

```mermaid
stateDiagram-v2
    direction LR
    [*] --> A
    state A {
        direction LR
        [*] --> B: a
        B --> C: b
        C --> [*]: c
    }
    A --> [*]
```

marcosh replied:
I do love Mermaid, but for these diagram I kind of prefer to be able to control precisely how things are represented, especially for more complex diagrams like this one:

![Screenshot_20230131_095000](https://user-images.githubusercontent.com/2643972/215712893-cbd7eaef-6f5c-4c51-a18d-1c4f6cdbba28.png)

If in the future we will use Mermaid to generate similar diagrams automatically, then it would definitely make sense to use the same tool in the documentation

AlexeiDrake reviewed (changes requested):
I'll hold off on reviewing this until the previous PRs are merged.

AlexeiDrake commented:
I'm really surprised at how big this PR has become. Why are there new decisions and code changes in the diff?

marcosh commented:
> I'm really surprised at how big this PR has become. Why are there new decisions and code changes in the diff?
> 
> Can we split this up into multiple PRs that all target `main`?

I rebased this PR on the of the latest developments, so that I could document those, too. It would probably be easier to review commit by commit. Otherwise, if you prefer, I could split the PR in several ones which will need to be reviewed one after the other

AlexeiDrake commented on spec/CRM/Example/RiskManager/Projection.hs:
```diff
@@ -0,0 +1,70 @@
+
+instance Semigroup ReceivedData where
+  (<>) :: ReceivedData -> ReceivedData -> ReceivedData
+  (<>) (ReceivedData ud1 ld1 cbd1) (ReceivedData ud2 ld2 cbd2) =
+    ReceivedData
+      (getLast $ Last ud1 <> Last ud2)
+      (getLast $ Last ld1 <> Last ld2)
+      (getLast $ Last cbd1 <> Last cbd2)
```
```suggestion
r1 <> r2 = ReceivedData
    { receivedUserData = receivedUserData r1 <|> receivedUserData r2
    , receivedLoanDetails = receivedLoanDetails r1 <|> receivedLoanDetails r2
    , receivedCreditBureauData = receivedCreditBureauData r1 <|> receivedCreditBureauData r2
    }
```

marcosh replied:
makes sense, but it still needs the `Last` and `getLast`

AlexeiDrake commented on spec/CRM/RiskManagerSpec.hs:
```diff
@@ -0,0 +1,191 @@
+
+creditBureauData :: CreditBureauData
+creditBureauData =
+  CreditBureauData
+    (MissedPaymentDeadlines 2)
+    (EuroCents 100000)
```
```suggestion
  CreditBureauData
    { missedPaymentDeadlines = MissedPaymentDeadlines 2
    , arrears = EuroCents 100000
    }
```

marcosh replied:
that's better, thanks

AlexeiDrake commented on docs/how-to-create-a-machine.md:
```diff
@@ -6,20 +6,22 @@ We will represent a machine as such
+- `Kleisli`
 
-Let's start with the last three.
+Let's start with the last five.
```
```suggestion
Every state machine that is not `Basic` is a composition of two state machines, which are ultimately expressed in terms of `Basic`.

Let's skip over `Basic` for now, and look at the other constructors first.
```

This is why I think `Compose` should be named `Sequence`; all of these constructors are compositions, but not every composition is sequential.

marcosh replied:
yes, now I agree with you. I'll do the change in a separate PR

AlexeiDrake commented on docs/how-to-create-a-machine.md:
````diff
@@ -127,6 +129,78 @@ where the `⊕` symbol allows us to distinguish alternative composition from the
+    └────┘
+```
+
+We can compose them like so
````
Case in point that `Compose` is a misleading constructor name. :)

AlexeiDrake commented on docs/how-to-create-a-machine.md:
```diff
@@ -127,6 +129,78 @@ where the `⊕` symbol allows us to distinguish alternative composition from the
+
+where the outputs `bs :: [b]` of the first machine are is passed as inputs to the second machine, and processed all one by one.
+
+Notice that in fact everything works not only for lists, but for every foldable type.
```
```suggestion
We use the `[b]` notation to denote multiple outputs, but we can in fact use any `Foldable` type.
```

AlexeiDrake commented on docs/how-to-create-a-machine.md:
```diff
@@ -165,3 +239,129 @@ In practice, a `Topology` is a list of pairs, where every pair contains a vertex
+
+To define a `Topology`, we first need to define the type of its vertices.
+
+This is usually done by defining an enumeration of the vertices, like
```
```suggestion
A straightforward approach is to create a sum type of vertices:
```

marcosh replied:
I also thought initially to say `sum type`, but also `Either Int String` is a sum type and that probably won't work here because then it won't satisfy the `Demote vertex ~ vertex` constraint we have in the `Basic` constructor for `StateMachine`.

In fact, I need to add some more information about those constraints and what they imply (basically they imply that the type must be finite)

AlexeiDrake commented on src/CRM/StateMachine.hs:
```diff
@@ -40,14 +42,22 @@ data StateMachine input output where
     -> StateMachine (Either a c) (Either b d)
+  Loop
+    :: StateMachine a [a]
+    -> StateMachine a [a]
```
I see this was done in a later commit. :)

marcosh commented on src/CRM/Graph.hs:
```diff
@@ -21,10 +23,30 @@ productGraph (Graph edges1) (Graph edges2) =
+      []
+      (nub $ fst <$> edges)
+  where
+    pathsStartingWith :: Eq a => Graph a -> a -> [(a, a)]
+    pathsStartingWith graph'@(Graph edges') a =
+      let
+        edgesStartingWithA = filter ((== a) . fst) edges'
+       in
+        edgesStartingWithA <> ((a,) . snd <$> (pathsStartingWith graph' . snd =<< edgesStartingWithA))
```
definitely more readable, thanks!

marcosh commented on src/CRM/BaseMachine.hs:
```diff
@@ -101,51 +124,73 @@ data InitialState (state :: vertex -> Type) where
+  -> ActionResult m topology state initialVertex output
+pureResult output state = ActionResult . pure $ (output, state)
+
+sequence
+  :: ActionResult Identity topology state initialVertex [output]
+  -> ActionResult [] topology state initialVertex output
+sequence (ActionResult (Identity (outputs, state))) =
+  ActionResult $ (,state) <$> outputs
```
this is actually not used, so I'll just remove it.

Anyway, I call it `sequence` because what it does is commuting two effects, like `sequence` does, but the type signature is not the same and actually not compatible

marcosh replied:
maybe this could actually be useful, so I'll keep it instead...

It could be generalised to

```haskell
sequence
  :: Functor f
  => ActionResult Identity topology state initialVertex (f output)
  -> ActionResult f topology state initialVertex output
```

On the other hand, `sequenceA` from `Traversable` would need to be

```haskell
sequence
  :: Functor f
  => ActionResult m topology state initialVertex (f output)
  -> f (ActionResult m topology state initialVertex output)
```

I'll add a comment to explain this

marcosh commented on src/CRM/BaseMachine.hs:
```diff
@@ -101,51 +124,73 @@ data InitialState (state :: vertex -> Type) where
+  :: (forall x. m x -> n x)
+  -> ActionResult m topology state initialVertex output
+  -> ActionResult n topology state initialVertex output
+hoistResult f (ActionResult outputStatePair) = ActionResult $ f outputStatePair
```
while with `fmap` you change with a function the type of your last parameter `output` in this case, with `hoist` you use a natural transformation (i.e. a parametric function, that's why the `forall` needs to be explicit) to change the context: in this case `m` becomes `n`.

It's the same idea you use in `Servant` to change context, see https://docs.servant.dev/en/stable/tutorial/Server.html#using-another-monad-for-your-handlers
